nflow-engine/src/main/java/io/nflow/engine/service/WorkflowInstanceService.java

```diff
@@ -90,23 +95,29 @@ public boolean updateWorkflowInstance(WorkflowInstance instance, WorkflowInstanc
     Assert.notNull(instance, "Workflow instance can not be null");
     Assert.notNull(action, "Workflow instance action can not be null");
     Assert.notNull(workflowDefinitionService, "workflowDefinitionService can not be null");
-
-    WorkflowInstance.Builder builder = new WorkflowInstance.Builder(instance);
-    if (instance.state == null) {
-      builder.setStatus(null);
-    } else {
-      String type = workflowInstanceDao.getWorkflowInstanceType(instance.id);
-      AbstractWorkflowDefinition<?> definition = workflowDefinitionService.getWorkflowDefinition(type);
-      builder.setStatus(definition.getState(instance.state).getType().getStatus(instance.nextActivation));
-    }
-    WorkflowInstance updatedInstance = builder.build();
-    boolean updated = workflowInstanceDao.updateNotRunningWorkflowInstance(updatedInstance);
-    if (updated) {
-      String currentState = workflowInstanceDao.getWorkflowInstanceState(updatedInstance.id);
-      WorkflowInstanceAction updatedAction = new WorkflowInstanceAction.Builder(action).setState(currentState).build();
-      workflowInstanceDao.insertWorkflowInstanceAction(updatedInstance, updatedAction);
+    try {
+      WorkflowInstance.Builder builder = new WorkflowInstance.Builder(instance);
+      if (instance.state == null) {
+        builder.setStatus(null);
+      } else {
+        String type = workflowInstanceDao.getWorkflowInstanceType(instance.id);
+        AbstractWorkflowDefinition<?> definition = workflowDefinitionService.getWorkflowDefinition(type);
+        builder.setStatus(definition.getState(instance.state).getType().getStatus(instance.nextActivation));
+      }
+      WorkflowInstance updatedInstance = builder.build();
+      boolean updated = workflowInstanceDao.updateNotRunningWorkflowInstance(updatedInstance);
+      if (updated) {
+        String currentState = workflowInstanceDao.getWorkflowInstanceState(updatedInstance.id);
+        WorkflowInstanceAction updatedAction = new WorkflowInstanceAction.Builder(action).setState(currentState).build();
+        workflowInstanceDao.insertWorkflowInstanceAction(updatedInstance, updatedAction);
+      } else {
+        // this is to trigger EmptyResultDataAccessException if instance does not exist
+        workflowInstanceDao.getWorkflowInstance(instance.id, emptySet(), 0L);
```

Comment:
Why not just `throw NflowNotFoundException("Workflow instance", instance.id, e)` directly?

Reply:
update can be false also when instance exists (but is executing and thus not updated)